include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -4418,6 +4421,666 @@ namespace internal
+                                   Number> &            fe_eval,
+        const bool                                      transpose,
+        const std::array<unsigned int, Number::size()> &c_mask,
+        Number *                                        values)
+    {
+      Assert(is_face == false, ExcInternalError());
+
+      if (transpose)
+        run_internal<fe_degree, true>(n_desired_components,
+                                      fe_eval,
+                                      c_mask,
+                                      values);
+      else
+        run_internal<fe_degree, false>(n_desired_components,
+                                       fe_eval,
+                                       c_mask,
+                                       values);
+
+      return false;
+    }
+
+  private:
+    template <int fe_degree, unsigned int side, bool transpose>
+    static void
+    interpolate_2D(const unsigned int given_degree,
+                   bool               is_subface_0,
+                   const unsigned int v,
+                   const Number *     weight,
+                   Number *           values)
+    {
+      if (is_subface_0)
+        interpolate_2D<fe_degree, side, transpose, true>(given_degree,
+                                                         v,
+                                                         weight,
+                                                         values);
+      else
+        interpolate_2D<fe_degree, side, transpose, false>(given_degree,
+                                                          v,
+                                                          weight,
+                                                          values);
+    }
+
+    template <int          fe_degree,
+              unsigned int side,
+              bool         transpose,
+              bool         is_subface_0>
+    static void
+    interpolate_2D(const unsigned int given_degree,
+                   const unsigned int v,
+                   const Number *     weight,
+                   Number *           values)
+    {
+      typename Number::value_type temp[10 /*TODO*/];
+
+      const unsigned int points =
+        (fe_degree != -1 ? fe_degree : given_degree) + 1;
+
+      const unsigned int d = side / 2; // direction
+      const unsigned int s = side % 2; // left or right surface
+
+      const unsigned int offset = dealii::Utilities::pow(points, d + 1);
+      const unsigned int stride =
+        (s == 0 ? 0 : (points - 1)) * dealii::Utilities::pow(points, d);
+
+      const unsigned int r1 = dealii::Utilities::pow(points, dim - d - 1);
+      const unsigned int r2 = dealii::Utilities::pow(points, d);
+
+      // copy result back
+      for (unsigned int i = 0, k = 0; i < r1; ++i)
+        for (unsigned int j = 0; j < r2; ++j, ++k)
+          temp[k] = values[i * offset + stride + j][v];
+
+      // perform interpolation point by point (note: r1 * r2 == points^(dim-1))
```

Comment:
But dim is 2 here, right? The reason I'm asking is that below you use `points - 1 - k` which only makes sense if `k` runs between 0 and `points`, so we don't support any other case here.

Reply:
Yes dim is 2 here. Essentially, the basis of this code is are the kernels from hyper.deal where we loop in d-dimensional space over all points of any face.